### DISCUSSION TITLE
Switch checks are ignored when the item interaction is cancelled?

### DISCUSSION THREAD

**Discussion by TheBusyBiscuit** [TownyAdvanced/Towny#4613] [Q&A] [answered]
Hey,

I have a question about the codebase of Towny.
So someone has recently submitted a bug report to us that cancelled item interactions bypass Towny protections ([Slimefun4#2641](https://github.com/Slimefun/Slimefun4/issues)).

We were able to track it down to us cancelling the item interaction by setting the ItemUse-Action to `DENY`.
But somehow this also disables the Towny check for container interactions, which seems quite strange to us...

Why exactly is this `!event.useItemInHand().equals(Event.Result.DENY))` in place there?
We can't really tell that from a pure look, as it isn't documented either.

https://github.com/TownyAdvanced/Towny/blob/dc29ece5a5022539b3697100d11274899cc5390c/src/com/palmergames/bukkit/towny/listeners/TownyPlayerListener.java#L317-L327

Coincidentally, `git blame` shows commit https://github.com/TownyAdvanced/Towny/commit/22a9b4c620dfff40d775441206fbc012f14a60fe to be the last change to this file, a commit that "supposedly" fixes the exact issue which appeared since then.

Would love to get some insight into why this Item Use check is in place, since item and block interactions should be rather distinct from each other.

cc: @WalshyDev

**Comment by LlmDl** [comment, marked as answer]
I want to say I added that in because it worked to stop the unregistered slimefun items bypassing the Towny switch use tests. Did something change on your guys' end since then where you're no longer setting that action to DENY useItemInHand?

**Reply by TheBusyBiscuit** [reply]
Until a few weeks ago, we weren't aware of this issue at all.
It only seemed to pop up recently and nothing changed on our end to my knowledge, although I could be wrong about that.

I tested it myself and were able to reproduce it, even debugged the state of our Listener and found out that only if the `useItemInHand` result is set to DENY this issue occurs, which is confirmed by the code snippet I linked.

And as far as I know, the `useItemInHand` Result was always set to DENY if the item was not allowed to be used.
That shouldn't be any new behaviour.

So I still don't really get why the DENY result would be excluded here, if this was in place just to fix this particular issue from us, then I think it could be removed as it seems to cause even more problems than it fixed.
Perhaps it was just a mistake on our end that got resolved by accident (happens regularly) and now the check is no longer needed?
Anyway, I don't really see any reason why the DENY `useItemInHand` result should be excluded from the protection checks here.

**Reply by LlmDl** [reply]
Yeah I'll give it a test/debug on my end and re-adjust things. A couple users brought this up in the discord yesterday which was a bit odd because when I'd last left it, it was fixed. Such is how it goes sometimes.

**Reply by TheBusyBiscuit** [reply]
Hmm, alright then! Thanks a lot :)

**Reply by LlmDl** [reply]
So the line `if (!event.useItemInHand().equals(Event.Result.DENY)) { ` wasn't the "fix" put in for slimefun last time, it was a block of code above it which solved for the slimefun/towny interaction.

When I re-jiggered things a little while back I nested the block of code which was the "fix" inside of the `if (!event.useItemInHand().equals(Event.Result.DENY)) { ` statement which re-opened the issue on Towny's side. 

The actual if statement's origin goes back further and isn't really needed at all any longer.
